### PR TITLE
Defer get_user_model calls to prevent circular imports

### PR DESCRIPTION
**Pull request by jctanner**
https://github.com/pulp/pulp_container/pull/1294#issuecomment-1602282563

### CODE REVIEW

File: pulpcore/app/role_util.py [hunk 1]
```diff
@@ -300,6 +298,7 @@ def get_users_with_perms_roles(
     include_model_permissions=True,
     for_concrete_model=False,
 ):
+    User = get_user_model()
```
Comment (lubosmj) [inline pulpcore/app/role_util.py:301]: I am not sure we can accept this as it is. Calling the function whenever working with permissions may not be the best approach. Can we cache `get_user_model` in a singleton class somewhere? More suggestions are welcome.

Comment (jctanner) [reply pulpcore/app/role_util.py:301]: I'm not following or understanding how a "singleton" would be implemented in this case. Is there another example of pulpcore where something like that has been done?

Comment (lubosmj) [reply pulpcore/app/role_util.py:301]: I mean to utilize the singleton pattern in general. Maybe, you could simply resort to `@lru_cache` if possible.

```
@lru_cache()
def blabla():
    return get_user_model()
```

**Comment by jctanner** [comment]
@lubosmj is this more what you were looking for?

**Comment by lubosmj** [comment]
Sort of. @jctanner, can you squash the commits, please?

**Comment by lubosmj** [comment]
Also, it is worth opening an issue and adding a changelog.

Perhaps, we want to set maxsize=1 in lru_cache as well if we decided to use it. Other folks may have different opinion on that. I still think caching is faster than doing always a dictionary and settings lookup. Can we rule out memory leaks once it will hold a reference to django_apps.Model?

**Comment by mdellweg** [comment]
Not sure how much overhead the lru_cache doing a dictionary lookup with no-args is bringing. Is the`get_users_with_perms_roles` hot enough to care?

File: pulpcore/app/models/role.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ class UserRole(BaseModel):
     """
 
     user = models.ForeignKey(
-        get_user_model(), related_name="object_roles", on_delete=models.CASCADE
+        settings.AUTH_USER_MODEL, related_name="object_roles", on_delete=models.CASCADE
```
Comment (mdellweg) [inline pulpcore/app/models/role.py:44]: :+1:

**Review by mdellweg — COMMENTED** [review]
Are we now properly breaking the dependenc loops in a way that we don't need the role_util to be it's own separate import path anymore?

OTOH, 500+ lines is probably a new reason for a separate module...

**Comment by lubosmj** [comment]
> Not sure how much overhead the lru_cache doing a dictionary lookup with no-args is bringing. Is the`get_users_with_perms_roles` hot enough to care?

It is not that hot. But, I believe these functions can be called from a ViewSet which could cause a tiny slowdown (unverified hypothesis).

**Comment by mdellweg** [comment]
Re changelog. We should have a bugfix in the plugin section for this (It effects a plugin quite seriously).

And also thanks for finding this much better solution than all the other moved imports!

**Review by gerrod3 — APPROVED** [review]

**Pull request merged**